packages/flutter/lib/src/widgets/selectable_text.dart

```diff
@@ -0,0 +1,602 @@
+  /// closest enclosing [DefaultTextStyle].
+  ///
+  /// The [data] parameter must not be null.
+  const SelectableText(
+    this.data, {
+    Key key,
+    this.focusNode,
+    this.style,
+    this.strutStyle,
+    this.textAlign,
+    this.textDirection,
+    this.showCursor = false,
+    this.autofocus = false,
+    this.maxLines,
+    this.cursorWidth = 2.0,
+    this.cursorRadius,
+    this.cursorColor,
+    this.dragStartBehavior = DragStartBehavior.start,
+    this.enableInteractiveSelection = true,
+    this.onTap,
+    this.scrollPhysics,
+    this.textWidthBasis,
+  }) :  assert(showCursor != null),
+        assert(autofocus != null),
+        assert(dragStartBehavior != null),
+        assert(maxLines == null || maxLines > 0),
+        assert(
+          data != null,
+          'A non-null String must be provided to a SelectableText widget.',
+        ),
+        textSpan = null,
+        super(key: key);
+
+  /// Creates a selectable text widget with a [TextSpan].
+  ///
+  /// The [textSpan] parameter must not be null.
+  const SelectableText.rich(
+    this.textSpan, {
+    Key key,
+    this.focusNode,
+    this.style,
+    this.strutStyle,
+    this.textAlign,
+    this.textDirection,
+    this.showCursor = false,
+    this.autofocus = false,
+    this.maxLines,
+    this.cursorWidth = 2.0,
+    this.cursorRadius,
+    this.cursorColor,
+    this.dragStartBehavior = DragStartBehavior.start,
+    this.enableInteractiveSelection = true,
+    this.onTap,
+    this.scrollPhysics,
+    this.textWidthBasis,
+  }) :  assert(showCursor != null),
+    assert(autofocus != null),
+    assert(dragStartBehavior != null),
+    assert(maxLines == null || maxLines > 0),
+    assert(
+      textSpan != null,
+      'A non-null TextSpan must be provided to a SelectableText.rich widget.',
+    ),
+    data = null,
+    super(key: key);
+
+  /// The text to display.
+  ///
+  /// This will be null if a [textSpan] is provided instead.
+  final String data;
+
+  /// The text to display as a [TextSpan].
+  ///
+  /// This will be null if [data] is provided instead.
+  final TextSpan textSpan;
+
+  /// Defines the focus for this widget.
+  ///
+  /// Text is only selectable when widget is focused.
+  ///
+  /// The [focusNode] is a long-lived object that's typically managed by a
+  /// [StatefulWidget] parent. See [FocusNode] for more information.
+  ///
+  /// To give the focus to this widget, provide a [focusNode] and then
+  /// use the current [FocusScope] to request the focus:
+  ///
+  /// ```dart
+  /// FocusScope.of(context).requestFocus(myFocusNode);
+  /// ```
+  ///
+  /// This happens automatically when the widget is tapped.
+  ///
+  /// To be notified when the widget gains or loses the focus, add a listener
+  /// to the [focusNode]:
+  ///
+  /// ```dart
+  /// focusNode.addListener(() { print(myFocusNode.hasFocus); });
+  /// ```
+  ///
+  /// If null, this widget will create its own [FocusNode].
+  final FocusNode focusNode;
```

Comment:
This makes sense to me but I want to pass it by @gspencergoog: Since I think you've been doing work with keyboard focus, does this fit with your idea of the focus system?  With SelectableText some text will be selectable by the user like on the web, and it can also be given a `focusNode`, even though it doesn't actually show a collapsed cursor (I don't think) and isn't editable.

Reply:
The main reason i need focus node is I still need to listen to keyboard event even tho we don't show a onscreen keyboard

Reply:
This seems like an OK use for `FocusNode` to me, although if you are just doing it to get the keyboard handling, then I don't see why the developer needs to be able to supply one.  You could just use a `Focus` widget internally, setting `onKey`, and not need to manage the focus node at all.